### PR TITLE
Support for named listeners

### PR DESCRIPTION
**Pull request by mhowlett**
Allow ApplicationServer to be instantiated with a list of named listeners, which will used in preference to the 'listener' configuration property if specified. Application instances can then be assigned to a specific listener by name or to all listeners.

Replaces https://github.com/confluentinc/rest-utils/pull/266 , which additionally contains various related functionality for parsing named listeners / protocol mapping in the configuration. That functionality has moved to ce-kafka-http-server.

I prefer the approach taken in this PR for two reasons:
1. It seems a bit wrong for Schema Registry, for example, to expose named listener functionality when there's absolutely no need for it. Even though it's optional.
2. The structure was not as clean as I guessed it might be (what initially lead me down that path) - the need to understand configuration was split between rest-utils and ce-kafka-http-server. Although I don't consider the structure here perfect (configuration of named listeners seems like it should sit side by side the code that makes it work), I think it is cleaner than the alternate approach.

### CODE REVIEW

File: core/src/main/java/io/confluent/rest/ApplicationServer.java [hunk 1]
```diff
@@ -104,79 +110,67 @@ public ApplicationServer(T config, ThreadPool threadPool) {
...
    * It's only used to support the deprecated configuration.
    */
   static List<URI> parseListeners(
-          List<String> listenersConfig,
+          List<String> listeners,
           int deprecatedPort,
           List<String> supportedSchemes,
-          String defaultScheme
-  ) {
+          String defaultScheme) {
```
Comment (mhowlett) [inline core/src/main/java/io/confluent/rest/ApplicationServer.java:124]: note: this method didn't really need to be modified, but at one point as i was implementing this it did, and i thought my changes represent an improvement, so left them.

**Review by rigelbm — APPROVED** [review]
LGTM

Approach looks good. Only nit comments.

File: core/src/main/java/io/confluent/rest/Application.java [hunk 2]
```diff
@@ -273,6 +282,13 @@ public Handler configureHandler() {
     ServletContextHandler context = new ServletContextHandler(ServletContextHandler.SESSIONS);
     context.setContextPath(path);
 
+    if (this.listenerName != null && !this.listenerName.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (mhowlett) [reply core/src/main/java/io/confluent/rest/Application.java:285]: Yeah ok. My preference is to always prefix with `this` when appropriate as it brings clarity, but i do see it's not the convention in this codebase. Java flaw mostly IMO :-).

File: core/src/main/java/io/confluent/rest/RestConfig.java [hunk 3]
```diff
@@ -1016,4 +1017,32 @@ public final boolean getDosFilterTrackGlobal() {
   public final boolean getDosFilterManagedAttr() {
     return getBoolean(DOS_FILTER_MANAGED_ATTR_CONFIG);
   }
+
+  public final Map<String,String> getMap(String propertyName, boolean keyValToLowerCase) {
```
[the remark this answers is not in the document]
Comment (mhowlett) [reply core/src/main/java/io/confluent/rest/RestConfig.java:1021]: yeah, good point, whoops. I assume you can do this as a streams one-liner. I'll get rid of this overload assuming that's easy.

File: core/src/test/java/io/confluent/rest/ApplicationServerTest.java [hunk 4]
```diff
@@ -116,9 +131,30 @@ protected ResourceCollection getStaticResources() {
     assertThat(makeGetRequest("/app2/index.html"), is(Code.OK));
   }
 
+  List<URL> getListeners() {
+    return Arrays.stream(server.getConnectors())
+            .filter(connector -> connector instanceof ServerConnector)
```
[the remark this answers is not in the document]
Comment (mhowlett) [reply core/src/test/java/io/confluent/rest/ApplicationServerTest.java:136]: in my defense, this and next are copy / paste thing.

File: core/src/main/java/io/confluent/rest/RestConfig.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mhowlett) [reply core/src/main/java/io/confluent/rest/RestConfig.java:1021]: oh, the duplicate key check makes that slightly interesting, but Collectors.toMap does that ootb.

File: core/src/main/java/io/confluent/rest/ApplicationServer.java [hunk 5]
```diff
@@ -87,11 +86,19 @@ public ApplicationServer(T config) {
...
 
+  public ApplicationServer(T config, List<NamedURI> namedListeners) {
+    this(config, namedListeners, createThreadPool(config));
+  }
+
   public ApplicationServer(T config, ThreadPool threadPool) {
+    this(config, null, threadPool);
+  }
+
+  public ApplicationServer(T config, List<NamedURI> namedListeners, ThreadPool threadPool) {
```
Comment (rigelbm) [inline core/src/main/java/io/confluent/rest/ApplicationServer.java:97]: I find it strange that we parse non-named listeners inside `ApplicationServer` but have to pass named-listeners as an argument. Why can't we parse named-listeners inside `ApplicationServer` as well?

File: core/src/main/java/io/confluent/rest/ApplicationServer.java [hunk 6]
```diff
@@ -386,7 +384,8 @@ SslContextFactory getSslContextFactory() {
     return this.sslContextFactory;
   }
 
-  private void configureConnectors(SslContextFactory sslContextFactory) {
+  private void configureConnectors(SslContextFactory sslContextFactory,
+                                   List<NamedURI> namedListeners) {
```
Comment (rigelbm) [inline core/src/main/java/io/confluent/rest/ApplicationServer.java:388]: This go back to the comment I made in the constructor. Here, we parse the non-named listeners inside the method, but named-listeners have to passed as an argument. I find it inconsistent.

**Review by rigelbm — CHANGES_REQUESTED** [review]
I thought I little bit more about it and add 2 (they are actually the same) comments that we need to discuss before merging.

File: core/src/main/java/io/confluent/rest/ApplicationServer.java (hunk 5 above)
Comment (mhowlett) [reply core/src/main/java/io/confluent/rest/ApplicationServer.java:97]: this was the approach i took initially in #266 (see the main PR comment). we could do it that way.

**Comment by mhowlett** [comment]
superseded by #268

**Pull request closed without merging**